// dynamic-core/esm2015/lib/dynamic-layout/BaseLayoutComponent.js
import { __awaiter } from "tslib";
import { ViewChildren, ChangeDetectorRef, Compiler, NgModuleFactory, Injector, Inject, Component } from '@angular/core';
import { tap } from 'rxjs/operators';
import { LayoutOutletDirective } from './layout-outlet.directive';
import { DashboardService } from './dynamic-layout.service';
import { DYNAMIC_MODULES_MAP } from './mapped-modules.interface';
import { ActivatedRoute } from '@angular/router';
export class BaseLayoutComponent {
    constructor(cd, injector, compiler, service, route, modulesMap) {
        this.cd = cd;
        this.injector = injector;
        this.compiler = compiler;
        this.service = service;
        this.route = route;
        this.modulesMap = modulesMap;
        this.containers = [];
        this.loadContents = () => {
            if (!this.dashboardOutlet || !this.dashboardOutlet.length) {
                return;
            }
            this.dashboardOutlet.forEach(template => {
                this.cd.detectChanges();
                this.loadContent(template, template.item);
            });
            this.cd.detectChanges();
        };
        this.loadContent = (template, item) => __awaiter(this, void 0, void 0, function* () {
            if (!item.name) {
                return;
            }
            const viewContainerRef = template.viewContainerRef;
            viewContainerRef.clear();
            const moduleOrFactory = yield this.modulesMap[item.name].load();
            let moduleFactory;
            if (moduleOrFactory instanceof NgModuleFactory) {
                moduleFactory = moduleOrFactory; // AOT
            }
            else {
                moduleFactory = yield this.compiler.compileModuleAsync(moduleOrFactory); // JIT
            }
            const moduleRef = moduleFactory.create(this.injector);
            console.log(item.cName);
            const rootComponent = moduleFactory.moduleType.entryComponents[`${item.cName}ContainerComponent`];
            const factory = moduleRef.componentFactoryResolver.resolveComponentFactory(rootComponent);
            const componentRef = viewContainerRef.createComponent(factory);
            const instance = componentRef.instance;
            instance.item = item;
        });
        this.changed = (items, trackIndex) => {
            const state = this.containers;
            state[trackIndex].items = items;
            this.service.setState(state);
        };
    }
    ngOnInit() {
        const id = this.route.snapshot.params['id'];
        this.service.loadContents(id).pipe(tap(tracks => (this.containers = tracks))
        /* Make sure to unsubscribe! */
        )
            .subscribe(() => {
            this.cd.detectChanges();
            this.loadContents();
        });
    }
    ngAfterViewInit() {
    }
}
BaseLayoutComponent.decorators = [
    { type: Component, args: [{
                template: ' '
            },] }
];
BaseLayoutComponent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: Injector },
    { type: Compiler },
    { type: DashboardService },
    { type: ActivatedRoute },
    { type: undefined, decorators: [{ type: Inject, args: [DYNAMIC_MODULES_MAP,] }] }
];
BaseLayoutComponent.propDecorators = {
    dashboardOutlet: [{ type: ViewChildren, args: [LayoutOutletDirective,] }]
};
