// dynamic-core/esm2015/lib/dynamic-layout/dynamic-layout.service.js
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { BaseSettingsService } from '../services/base-settings.service';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "./index";
import * as i2 from "../services/base-settings.service";
export class DashboardService {
    constructor(settingsService) {
        this.settingsService = settingsService;
        this.historyState = [];
        this.subject = new BehaviorSubject(this.historyState);
        this.setState = (tracks) => {
            this.subject.next(tracks);
        };
        this.addItem = (item) => {
            const state = this.subject.getValue();
            if (state[0].items.indexOf(item) !== -1 || state[1].items.indexOf(item) !== -1) {
                console.warn('Item with the same id exists on the dashboard.');
                return;
            }
            state[0].items.length <= state[1].items.length ? state[0].items.push(item) : state[1].items.push(item);
            this.subject.next(state);
        };
        this.removeItem = (item) => {
            const state = this.subject.getValue();
            state.forEach(track => {
                track.items.forEach((i, index) => {
                    if (i === item) {
                        track.items.splice(index, 1);
                    }
                });
            });
            this.subject.next(state);
        };
        this.loadContents = (id) => {
            return this.settingsService.loadSettings(id).pipe(map((state) => {
                this.historyState = state;
                this.subject.next(this.historyState);
                return state;
            }));
        };
    }
}
DashboardService.ɵprov = i0.ɵɵdefineInjectable({ factory: function DashboardService_Factory() { return new i1.DashboardService(i0.ɵɵinject(i2.BaseSettingsService)); }, token: i1.DashboardService, providedIn: "root" });
DashboardService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
DashboardService.ctorParameters = () => [
    { type: BaseSettingsService }
];
