// primeng-dynamic-core/fesm2015/primeng-dynamic-core.js
import { CommonModule } from '@angular/common';
import { Component, ChangeDetectorRef, Injector, Compiler, Inject, NgModule } from '@angular/core';
import { BaseLayoutComponent, DashboardService, DYNAMIC_MODULES_MAP, DynamicLayoutModule } from 'dynamic-core';
import { TabViewModule } from 'primeng/tabview';
import { ActivatedRoute } from '@angular/router';
import { PanelModule } from 'primeng/panel';

class TabLayoutComponent extends BaseLayoutComponent {
    constructor(cd, injector, compiler, dashboardService, route, modulesMap) {
        super(cd, injector, compiler, dashboardService, route, modulesMap);
    }
}
TabLayoutComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-tab-layout',
                template: "<div class=\"p-grid\">\n  <div class=\"p-col\" *ngFor=\"let container of containers\">\n    <p-tabView>\n      <p-tabPanel [header]=\"item.title\" *ngFor=\"let item of container.items; let i = index\" [selected]=\"i == 0\">\n        <ng-template appLayouOutlet [item]=\"item\"></ng-template>\n      </p-tabPanel>\n    </p-tabView>\n  </div>\n</div>\n",
                styles: [""]
            },] }
];
TabLayoutComponent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: Injector },
    { type: Compiler },
    { type: DashboardService },
    { type: ActivatedRoute },
    { type: undefined, decorators: [{ type: Inject, args: [DYNAMIC_MODULES_MAP,] }] }
];

class PanelLayoutComponent extends BaseLayoutComponent {
    constructor(cd, injector, compiler, dashboardService, route, modulesMap) {
        super(cd, injector, compiler, dashboardService, route, modulesMap);
    }
}
PanelLayoutComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-panel-layout',
                template: "<div class=\"p-grid\">\n  <div class=\"p-col\" *ngFor=\"let container of containers\">\n      <p-panel [header]=\"item.title\" *ngFor=\"let item of container.items;\">\n        <ng-template appLayouOutlet [item]=\"item\"></ng-template>\n      </p-panel>\n  </div>\n</div>",
                styles: [""]
            },] }
];
PanelLayoutComponent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: Injector },
    { type: Compiler },
    { type: DashboardService },
    { type: ActivatedRoute },
    { type: undefined, decorators: [{ type: Inject, args: [DYNAMIC_MODULES_MAP,] }] }
];

class PrimeDynamicCoreModule {
}
PrimeDynamicCoreModule.decorators = [
    { type: NgModule, args: [{
                declarations: [TabLayoutComponent, PanelLayoutComponent],
                imports: [
                    CommonModule, DynamicLayoutModule, TabViewModule, PanelModule
                ],
                exports: [TabLayoutComponent, PanelLayoutComponent, DynamicLayoutModule]
            },] }
];

/*
 * Public API Surface of prime-dynamic-core
 */

/**
 * Generated bundle index. Do not edit.
 */

export { PanelLayoutComponent, PrimeDynamicCoreModule, TabLayoutComponent };
