// primeng-dynamic-core/bundles/primeng-dynamic-core.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/core'), require('dynamic-core'), require('primeng/tabview'), require('tslib'), require('@angular/router'), require('primeng/panel')) :
    typeof define === 'function' && define.amd ? define('primeng-dynamic-core', ['exports', '@angular/common', '@angular/core', 'dynamic-core', 'primeng/tabview', 'tslib', '@angular/router', 'primeng/panel'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['primeng-dynamic-core'] = {}, global.ng.common, global.ng.core, global.dynamicCore, global.tabview, global.tslib, global.ng.router, global.panel));
}(this, (function (exports, common, core, dynamicCore, tabview, tslib, router, panel) { 'use strict';

    var TabLayoutComponent = /** @class */ (function (_super) {
        tslib.__extends(TabLayoutComponent, _super);
        function TabLayoutComponent(cd, injector, compiler, dashboardService, route, modulesMap) {
            return _super.call(this, cd, injector, compiler, dashboardService, route, modulesMap) || this;
        }
        return TabLayoutComponent;
    }(dynamicCore.BaseLayoutComponent));
    TabLayoutComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'app-tab-layout',
                    template: "<div class=\"p-grid\">\n  <div class=\"p-col\" *ngFor=\"let container of containers\">\n    <p-tabView>\n      <p-tabPanel [header]=\"item.title\" *ngFor=\"let item of container.items; let i = index\" [selected]=\"i == 0\">\n        <ng-template appLayouOutlet [item]=\"item\"></ng-template>\n      </p-tabPanel>\n    </p-tabView>\n  </div>\n</div>\n",
                    styles: [""]
                },] }
    ];
    TabLayoutComponent.ctorParameters = function () { return [
        { type: core.ChangeDetectorRef },
        { type: core.Injector },
        { type: core.Compiler },
        { type: dynamicCore.DashboardService },
        { type: router.ActivatedRoute },
        { type: undefined, decorators: [{ type: core.Inject, args: [dynamicCore.DYNAMIC_MODULES_MAP,] }] }
    ]; };

    var PanelLayoutComponent = /** @class */ (function (_super) {
        tslib.__extends(PanelLayoutComponent, _super);
        function PanelLayoutComponent(cd, injector, compiler, dashboardService, route, modulesMap) {
            return _super.call(this, cd, injector, compiler, dashboardService, route, modulesMap) || this;
        }
        return PanelLayoutComponent;
    }(dynamicCore.BaseLayoutComponent));
    PanelLayoutComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'app-panel-layout',
                    template: "<div class=\"p-grid\">\n  <div class=\"p-col\" *ngFor=\"let container of containers\">\n      <p-panel [header]=\"item.title\" *ngFor=\"let item of container.items;\">\n        <ng-template appLayouOutlet [item]=\"item\"></ng-template>\n      </p-panel>\n  </div>\n</div>",
                    styles: [""]
                },] }
    ];
    PanelLayoutComponent.ctorParameters = function () { return [
        { type: core.ChangeDetectorRef },
        { type: core.Injector },
        { type: core.Compiler },
        { type: dynamicCore.DashboardService },
        { type: router.ActivatedRoute },
        { type: undefined, decorators: [{ type: core.Inject, args: [dynamicCore.DYNAMIC_MODULES_MAP,] }] }
    ]; };

    var PrimeDynamicCoreModule = /** @class */ (function () {
        function PrimeDynamicCoreModule() {
        }
        return PrimeDynamicCoreModule;
    }());
    PrimeDynamicCoreModule.decorators = [
        { type: core.NgModule, args: [{
                    declarations: [TabLayoutComponent, PanelLayoutComponent],
                    imports: [
                        common.CommonModule, dynamicCore.DynamicLayoutModule, tabview.TabViewModule, panel.PanelModule
                    ],
                    exports: [TabLayoutComponent, PanelLayoutComponent, dynamicCore.DynamicLayoutModule]
                },] }
    ];

    /*
     * Public API Surface of prime-dynamic-core
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.PanelLayoutComponent = PanelLayoutComponent;
    exports.PrimeDynamicCoreModule = PrimeDynamicCoreModule;
    exports.TabLayoutComponent = TabLayoutComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
